### PR TITLE
[IMP] website: add Instagram page snippet

### PR DESCRIPTION
**Pull request by Guillaume-gdi**
This commit adds a new snippet to display an Instagram page on the website using an iframe that embeds the Instagram page.

task-2603045

### CODE REVIEW

File: addons/website/static/src/snippets/s_instagram_page/000.scss
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_instagram_page/000.scss:5]: I do this because I know we don't like inline css but here it might be worth putting this little rule inline in order to have one less asset ? :thinking:

**Comment by Guillaume-gdi** [comment]
Note that we have to wait for the thumbnail for this new block (task-3495454) :framed_picture:

**Review by robinlej — COMMENTED** [review]
Here's my pair review with some general comments:
- This should somehow be linked to the cookies acceptation. Have a fallback UI if third-party cookies haven't been accepted along the lines of "Sorry, to see this you need to accept third-party cookies", or even just hide the snippet altogether. And of course, don't make the call to Insta's API if we don't have the cookie approval.
- Considering the name of the option's input is "Instagram Page", I'm 100% sure some people will copy/paste the whole URL "https://www.instagram.com/account/" and expect it to work. I would add a `cleanPageName()` method to only keep the account in such cases.
- The previews made by the options feel sluggish because it's fetching the whole Instagram iframe each time it computes a preview, although except if you enter another account name, the option should only affect the section (its height / its background color / whatever, but the iframe stays the same). Since circumventing the preview reloading the whole snippet is probably non-trivial to say the least, maybe the best solution would be to keep a cache of the iframe as long as the editor is open and the snippet exists. What do you think?
- Whether or not you find a way to make it so that previews don't trigger a rerendering of the iframe, in either case, you could already add a loader in the place of the iframe while the server waits for the API's response. It already takes 1 or 2s with our Wifi, I can't imagine how long it would take on a poor connection. :grimacing:
- If you add a background shape, the Insta iframe disappears and the shape completely bugs (you can't select another one, can't change the color, can't remove it...). I really don't know why, but it looks like it's working on other snippets on the runbot, so it seems linked to your PR. :thinking:
- If I add a video as a background, the height of the iframe is slightly cropped (you can see it with the "View full profile" button, it's cropped).
- Not sure what's the best to deal with this, but if you conditionnally hide the snippet (say visible only when logged in), considering how the conditional visibility works, the embed is still loaded when you're logged out. Since this takes a long time and takes resources, I guess it would make sense not to load it in such cases. Maybe just checking whether the iframe `isVisible()` (from the web ui utils `addons/web/static/src/core/utils/ui.js`) before doing any of its magic would do the trick?
- I guess you've seen it, but there's an error in the console linked to Instagram's API. Any idea what this is?

File: addons/website/static/src/snippets/s_instagram_page/000.js
Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_instagram_page/000.js:58]: `a least-squares regression is used to estimate the height of the block.`
So this is not clear enough I assume :grimacing: . I will improve this part :+1:

**Comment by Guillaume-gdi** [comment]
Thanks for the good review @robinlej :+1: 

> This should somehow be linked to the cookies acceptation...

For the moment we can ignore it. We will find a solution in the [task-3027209](https://www.odoo.com/web#id=3027209&cids=1&menu_id=4720&action=333&active_id=974&model=project.task&view_type=form)

> Some people will copy/paste the whole URL

Indeed it's a quick win, it's adapted

> The previews made by the options feel sluggish because it's fetching the whole Instagram iframe each time

I fixed it by preventing the public widget to restart every time an option is applied.

> add a loader in the place of the iframe while the server waits for the API's response

This is the same issue with other's dynamic content (eg: products block). We will take care of that in [task-3496216](https://www.odoo.com/web#id=3496216&cids=1&menu_id=4720&action=333&active_id=974&model=project.task&view_type=form)

> Shape issue

Fixed :heavy_check_mark: 

> Background video issue

Fixed :heavy_check_mark: 

> Don't fetch instagram if the block is invisible

Done!

>  there's an error in the console linked to Instagram's API

That's strange.. When I go to https://www.instagram.com/odoo.official/embed/ or https://www.instagram.com/popsci/embed/ I don't have the console.error but on https://www.instagram.com/nike/embed/ I have it :raised_eyebrow: .. It's quite difficult to reverse engineer the JS that cause the error as the code is minified and obfuscated. I tried to add args in the url (the one I found on other instagram embed) but I can't find the way to solve it.. IMO it's an issue in the instagram code and we can live with that for now

**Comment by Guillaume-gdi** [comment]
@pierre-pde For now I put the Instagram block after the facebook block in dynamic content blocks group. Let me know if you want it elsewhere :+1: 
image

File: addons/website/static/src/snippets/s_instagram_page/000.js
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_instagram_page/000.js:96]: I created a task to fix this issue of public widget not destroyed on undo.
[Task-3508448](https://www.odoo.com/web#id=3508448&cids=1&menu_id=4720&action=333&active_id=974&model=project.task&view_type=form)

**Comment by Guillaume-gdi** [comment]
@qsm-odoo or @rdeodoo. As we want this for v17, It would be great to start the review process for this task when you have time :+1:

File: addons/website/static/src/snippets/s_instagram_page/000.js
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_instagram_page/000.js:97]: Don't parse multiple times

File: addons/website/static/src/snippets/s_instagram_page/000.js
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_instagram_page/000.js:112]: Maybe one line with `this.options.wysiwyg?.odooEditor.observerUnactive();`

File: addons/website/static/src/snippets/s_instagram_page/options.js
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_instagram_page/options.js:12]: You call that method only once, do not make a function out of it (and no need to make it overriddable)

File: addons/website/static/src/snippets/s_instagram_page/options.js
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_instagram_page/options.js:43]: This an async operation you should await (as done where you disable it)

File: addons/website/static/src/snippets/s_instagram_page/options.js
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_instagram_page/options.js:119]: Should not we rather export the class directly?

```
export default InstagramPage = options.class.blablabla;
options.registry.InstagramPage = InstagramPage;
```
Something like that?

File: addons/website/views/snippets/snippets.xml
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml:1236]: Why? :open_mouth:

File: addons/website/static/src/snippets/s_instagram_page/000.scss
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_instagram_page/000.scss:8]: Maybe add w-100 on the iframe when creating it to avoid this extra CSS rule?

File: addons/website/static/src/snippets/s_instagram_page/000.scss
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_instagram_page/000.scss:5]: Is a container really needed if we disable its only purpose? Are we doing this in any other snippet?

File: addons/website/static/src/js/editor/snippets.options.js
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:940]: Not a bad idea for now but add a TODO comment to say we should review the way public widget are restarted when converting to OWL and a new API.

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
A first quick review.

File: addons/website/static/src/snippets/s_instagram_page/000.js
Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_instagram_page/000.js:19]: >No jquery in new code?

I thought that for `:visible`, it was accepted as long as we don't have an alternative.

> Is this really needed anyway?

It's not essential.

>  if we need that, should not that be generic to all snippets?

Indeed yes, we could prevent public widget of invisible blocks to start. Do you want me to create a task for that (low priority) ?

=> I will remove the code from this PR :+1:

Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_instagram_page/000.js:19]: > accepted as long as we don't have an alternative.

Indeed :+1: 

> Do you want me to create a task for that (low priority) ?

Let's do that :+1:

File: addons/website/static/src/snippets/s_instagram_page/000.js
Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_instagram_page/000.js:94]: I just checked now on master and the issue is present. I checked the commit you mention and the problem is already there before. But indeed, it seems critical and we should add a test for it.

As I mentioned above I already created a [task-3508448](https://www.odoo.com/web#id=3508448&cids=1&menu_id=4720&action=333&active_id=974&model=project.task&view_type=form) to fix this issue.

Let me know if you want me to investigate it more before going forward on v17 tasks.

File: addons/website/views/snippets/snippets.xml
Comment (Guillaume-gdi) [reply addons/website/views/snippets/snippets.xml:1236]: image

File: addons/website/static/src/snippets/s_instagram_page/000.js
Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_instagram_page/000.js:94]: > Let me know if you want me to investigate it more before going forward on v17 tasks.

No, of course priority on task for now

File: addons/website/views/snippets/snippets.xml
Comment (qsm-odoo) [reply addons/website/views/snippets/snippets.xml:1236]: Oh, yes. We still have that bad UX where you can only have a scroll button if you set up a full height... sad.

File: addons/website/static/src/snippets/s_instagram_page/000.js
Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_instagram_page/000.js:19]: task created [here](https://www.odoo.com/web#id=3526079&menu_id=4720&cids=1&action=333&active_id=974&model=project.task&view_type=form)

File: addons/website/static/src/snippets/s_instagram_page/options.js
Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_instagram_page/options.js:12]: Sad, I think it really improves the readability... Can I do a function that is declared inside the onBuilt function ?

Comment (rdeodoo) [reply addons/website/static/src/snippets/s_instagram_page/options.js:12]: Randomly passing by.
By no mean do I think my opinion is the best one or that I am right, but here it is:

It's imo always a bad practice to have a "only call once" method to encapsulate some code.
It's also a bad practice to make a method which is only doing a single line operation even if called multiple times (unless that line is extremely difficult eg).

Otherwise, you enter the labyrinthe effect, you need to scroll down, scroll up, scroll down, scroll up, to read the code.
That's, IMHO, the first bad part.
The second bad part is that, the method name will add even more complexity unless you find the very perfect method name, which you won't because there is never one.
People will not get exactly the method name as you meant it to be, it will add a layer of confusion.

The same apply to variables, it's why when we can omit it, it's better.

This:
```py
def main():
   some_setup_basic_instruction()

   # Retrieve the XYZ
    instruction_1()
    instruction_2()
  
    # Adapt ABC
    instruction_1()
    while ():
        instruction_2()
    
    # Colour DEF
    instruction_1()
    if blabla:
        instruction_2()
```
Will imo always be better than
```py
def main():
   some_setup_basic_instruction()
   self.retrieve_xyz()
   self.adapt_abc()
   self.colour_def()

def retrieve_xyz()
    instruction_1()
    instruction_2()
  
def adapt_abc():
    instruction_1()
    while ():
        instruction_2()
    
def colour_def():
    instruction_1()
    if blabla:
        instruction_2()
```
Same for sub method
```py
def main():
    def retrieve_xyz()
        instruction_1()
        instruction_2()
  
    def adapt_abc():
        instruction_1()
        while ():
            instruction_2()
    
    def colour_def():
        instruction_1()
        if blabla:
            instruction_2()
          
   some_setup_basic_instruction()
   self.retrieve_xyz()
   self.adapt_abc()
   self.colour_def()
```

Because the first one just reads like a storyline, from A to Z, it's just like a simple flow.
And with good line return and comments, you can very well divide you code in clear sections, which is at the end what you are trying to accomplish by encapsulating it in variables.

It's exactly like the website generator PR, there was plenty of those "only called once" method or "short util" which made it impossible to read to me, I had to scroll to a sub method, to then scroll to another sub method, to then scroll to an util, to then scroll up the other way around.

Again, just my opinion.
But it's generally the first thing I ask to change.
`https://github.com/odoo/enterprise/pull/43608#discussion_r1298441773`
`https://github.com/odoo/odoo/pull/126719#discussion_r1322974772` (even if I was wrong and actually couldn't apply this in this case)

Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_instagram_page/options.js:12]: > It's also a bad practice to make a method which is only doing a single line operation even if called multiple times (unless that line is extremely difficult eg).

I agree with this and with the fact that we have to prevent to have too much deep in the call stack but the rest of your opinion seems to be against clean code principles. I read everywhere that functions should have only one responsibility and be small (max 30 lines).

File: addons/website/static/src/snippets/s_instagram_page/options.js
Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_instagram_page/options.js:119]: We can yes but it would be strange no, it would be the only option exported this way ?

File: addons/website/static/src/snippets/s_instagram_page/options.js
Comment (rdeodoo) [reply addons/website/static/src/snippets/s_instagram_page/options.js:12]: I really disagree with that (as suggested by my previous message :innocent: ), maybe it's because Odoo already has well defined and well separated method, hooks etc.

I don't see the point to divide again a `start`, `onBuilt` js method or a `create`, `name_get` python method for example.

It's just that it turns into labyrinth nightmares, amplified by method and variable bad names (which are required when dividing into methods) most of the times, since if you "allow" that, people will overdo it.

Really just a few days ago:
- `https://github.com/odoo/odoo/pull/126719#discussion_r1321589230` this was making me have to scroll to see what `ref()` was, then go back the method, then the day after, I couldn't remember if it was an alias to `env.ref()` and/or if it was crashing in case of not found, has to scroll again
- `https://github.com/odoo/odoo/pull/126719#discussion_r1323135889` -> what does `_adjust_new_page_template` mean? You have to scroll to figure, the method name doesn't hint anything (it surely did to the writer but not to me)

I really don't think it brought added value to encapsulate that code.

The thing is that sadly, we have to go for "never" or "always" regarding this case, because people will overdo it every single time they can if you say something like "Let's not do it unless we judge it's really worth doing".
That's what "clean code principles" usually tell: "It's always a case by case, if you judge blabla, then do blabla, it should be obvious blabla, there is no 'always' or 'never', blabla".
But in real world, working with many people on many files, **you have to** instore a 'always' or "never". People don't even care about having 2 typos and 2 linter errors by line, they don't care about blaming line they are removing, they just won't care about making good usage of the method splitting, the easiest and fastest, the better. And thinking case by case is not the easiest so they won't.

But even if we did it (allow case by case), in this case, it would not be a good use case to split the method IMHO.
It's very very very short and the code which is left in `start()` is actually bound to the one encapsulated in  `_getDbInstagramPageName`. So you would actually need to move the code left in `start()` in the sub method which would be named something like `_setInstagramPage` which would get the info and set the attribute: encapsulate the whole behavior / principle. No?

Also, quickly googling, I read this from a random guy (so it's worth what it's worth) which is "quoting" a book:
> Also, please note that contrary to what many are saying here, **there appears to be ample evidence that moderately long functions (say 50 to 200 lines) actually have lower bug densities than lots of very short functions**. McConnell provides a bunch of references in his book (although I've seen those references disputed). Again, as stated above, there's always an element of judgment when determining what belongs in a function.

It might be worth considering too. Using variables (to hold submethod return value) (on top of other things) can surely lead to more bugs.

If you are strongly against it, I am sure @qsm-odoo will let it go if it's something you really care about / want.
For me, I was just passing by discussing this hot topic :smiley:

**Comment by qsm-odoo** [comment]
@robodoo delegate=bvr-odoo 
@bvr-odoo

File: addons/website/static/src/snippets/s_instagram_page/options.js
Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_instagram_page/options.js:12]: We discussed it IRL, I will adapt my code to RDE/QSM comment :+1:

Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_instagram_page/options.js]: Some notes to justify why I am asking that: I personally was more inclined to create one-call functions if it seemed to make sense before. Then I spent years now reviewing code... and not once did it make my life easier to have a function called only once. For the many reasons RDE explained:
- The name of the function was not clear to me
- I have to needlessly jump in the code
- etc

In Lou's PR for ChatGPT, there were like 10 occurences of those one-call functions... reuniting stuff was the only way for me to understand the uncommented code. Doing that also fixed bugs.

That's why I changed my mind over the years :wink: But if you prefer like you did before, I don't mind. Don't get me wrong: one-call functions can make sense to me. That's basically different interpretations of https://en.wikipedia.org/wiki/Single-responsibility_principle . It can also make sense for unit testing, possibility to override, etc. And I don't think coding principles are necessarily always right: e.g. I discovered OOP by myself as a teenager, studied it at school later, it was as much as a coding principle as a way of programming... and today, it's really not as popular as it was before from my point of view (OWL proving it, although it has some OOP aspects).

File: addons/website/static/src/snippets/s_instagram_page/options.js
Comment (bvr-odoo) [inline addons/website/static/src/snippets/s_instagram_page/options.js:94]: Couldn't you to avoid repeating `"instagram.com/"` ? For example put it in a variable..

File: addons/website/views/snippets/s_instagram_page.xml
Comment (bvr-odoo) [inline addons/website/views/snippets/s_instagram_page.xml:35]: I guess you forgot to replace the name after copy/pasting here ;)

**Review by bvr-odoo — CHANGES_REQUESTED** [review]

File: addons/website/static/src/js/editor/snippets.options.js
Comment (Guillaume-gdi) [reply addons/website/static/src/js/editor/snippets.options.js:929]: I think we try to prevent adding a class to the snippet if we can. As this is useless outside of edit mode, I would keep it as it is.

**Review by bvr-odoo — APPROVED** [review]
Cool feature, thanks for your work :+1:

**Comment by bvr-odoo** [comment]
@robodoo r+

**Pull request closed without merging**
